boa/test/strategies.py

```diff
@@ -100,11 +101,14 @@ def generate_random_string(n):
 @_exclude_filter
 def _address_strategy(length: Optional[int] = 100) -> SearchStrategy:
     random_strings = generate_random_string(length)
-    # TODO: add addresses from the environment. probably everything in
-    # boa.env._contracts, boa.env._blueprints and boa.env.eoa.
     accounts = [format_addr(i) for i in random_strings]
+
+    # inject env addresses, ultimately, will sample length + 1 + len(boa.env._contracts) addresses
+    accounts += [boa.env.eoa]
+    accounts += [contract.address for contract in boa.env._contracts.values()]
+    # TODO: add all addresses generated by env through env.generate_address()
```

Comment:
- I don't think it's very likely, but it seems more correct to put the addresses in a `set`. Otherwise, we don't need to call `list` in line 111 - `accounts` is already a list
- The TODO should be removed
- Why aren't we including blueprints like decribed in the TODO?

```suggestion
    accounts = {format_addr(i) for i in random_strings} & {boa.env.eoa} & \
        {contract.address for contract in boa.env._contracts.values()} & \
        {blueprint.address for blueprint in boa.env._blueprints.values()}
```

Reply:
- The TODO is a new TODO. We discussed with @charles-cooper to maybe buffer all addresses generated by env.generate_address() in a new env attribute and include them in the address strategy. Thoughts?
- re blueprints, i do not see a _blueprints attribute in the env class. Was it replaced by _code_registry?

Reply:
`env._code_registry` includes both blueprints and contracts
saving all addresses seems like a good idea!

Reply:
shall i make a new issue for saving all generated addresses? Or use this one? @charles-cooper suggested to use [weakref.WeakSet](https://docs.python.org/3/library/weakref.html#weakref.WeakSet) to buffer generated addresses.

something like:

```
def generate_address(self, alias: Optional[str] = None) -> _AddressType:
        t = Address(self._random.randbytes(20))
        if alias is not None:
            self.alias(t, alias)

        # add to weakref.weakset _addresses
        self._addresses.add(t)
        return t
```

Reply:
The strategy would become

```
@_exclude_filter
def _address_strategy(length: Optional[int] = 100) -> SearchStrategy:
    random_strings = generate_random_string(length)

    accounts = ({format_addr(i) for i in random_strings} |
                {boa.env._addresses} |
                {obj.address for obj in boa.env._code_registry.values()})

    return _DeferredStrategyRepr(
        lambda: st.sampled_from(list(accounts)), "accounts"
    )
```

Reply:
That looks good! Maybe we could generate less random strings depending on how many addresses we have in the env.

Reply:
unfortunately, basic data types cannot be weakly referenced (including str), and Address class extends the str class. Do we need Address to extend str? Is it actually used? Can we add some methods to Address to be directly comparable to str without extending str?

Reply:
maybe we could use a custom overflowing set, but that s not so elegant.

Reply:
(the idea is to avoid having the memory footprint of the env getting out of control as more generated addresses are buffered)

Reply:
We could use the existing [`lrudict`](https://github.com/vyperlang/titanoboa/blob/1bf16f916b91aab299e293b4148be455c1674706/boa/util/lrudict.py) or a similar structure for sets to keep the memory footprint limited.

The weakset doesn't make sense indeed for simple types as they always get copied and never referenced.

Reply:
> opted in to use lrudict to keep a cache of latest generated addresses, to use in the address strategy

hang on, i think this should be out of scope as it requires a little more thought. can we revert this change? i thought it was fine as an initial change before

Reply:
ok i ll keep a TODO in the strat then

Reply:
> unfortunately, basic data types cannot be weakly referenced (including str), and Address class extends the str class. Do we need Address to extend str? Is it actually used? Can we add some methods to Address to be directly comparable to str without extending str?

yes, we do, because it compares to string cleanly in user tests. let's revisit it later then

Reply:
> That looks good! Maybe we could generate less random strings depending on how many addresses we have in the env.

no, let's keep generating random strings - we don't want to reduce fuzzing strength just bc a user already has addresses in the env

Reply:
> `env._code_registry` includes both blueprints and contracts saving all addresses seems like a good idea!

i was a little hesitant about this because blueprints can't really do things like make contract calls. but maybe that's actually a good thing? because then the user has to deal with addresses which contain un-executable code.

Reply:
btw, i saw that the random addresses are all like "0x0000000000000000000whatever", not covering the complete "bytespace" at all

Reply:
> btw, i saw that the random addresses are all like "0x0000000000000000000whatever", not covering the complete "bytespace" at all

yea that's busted, it's the existing definition for `generate_random_string`:
https://github.com/theodorecurtil/titanoboa/blob/1b54ccfc3f5690ee4b6cb6f6a1cfbd94cad47188/boa/test/strategies.py#L98C21-L98C35

honestly we should probably switch to something like https://docs.python.org/3/library/random.html#random.randbytes